Feature/10 notification

tetetearraio916 opened the pull request:
# 通知機能の実装
## 内容
- notificationsテーブルを作成。
- userテーブルに関連付け
- likes、comments、ralationshipsテーブルにポリモーフィックで関連付け
- 通知一覧をヘッダー用とプロフィール用でルーティングを場合わけ
- ヘッダーの通知は、bootstrapのdropdownで見れるように実装

## だいそんさんとのコードの違い
通知機能のテーブル名がactivityのところ、僕の場合notificationとなっています。

## 問題点
ドロップダウン機能が1クリックで開いたままにしたいが、一度開いてすぐに閉じてしまう。

tetetearraio916 committed:
create notification model

notificationモデル、マイグレーションファイルを作成しました。また、like、comment、
relationshipとポリモーフィック関連付けを行いました。

tetetearraio916 committed:
create_notifications method for post comment relationship

post、comment、relationshipの各モデルにcreate_notificationsメソッドを実装。
各モデルのレコードが保存されると発動するようにしています。
また、notificationのaction_type、checkedカラムに対してenumを使って名前定義しました。

tetetearraio916 committed:
create notification partial of like comment follow

notificationのパーシャルを作成しました。like、comment、followの通知の種類による
パーシャルも作成。

tetetearraio916 committed:
add notification routhing

notificationのルーティングを設定しました。
プロフィール一覧とヘッダー一覧用の二つのルーティングを設定しました。

tetetearraio916 committed:
create notification controller

notificationのコントローラを実装しました。

tetetearraio916 committed:
add redirect_path method in notification model

notificationnモデルにredirect_pathを追加しました。
like、comment、followの通知の種類によってリダイレクト先を分けるメソッドです。
notificationコントローラで使用しています。

tetetearraio916 committed:
add notification for header

ヘッダーに通知機能を実装しました。
問題点が残っています。
クリックすると通知機能が一瞬しか開けない点です。
1クリックで開くと閉じる実行されている。

tetetearraio916 committed:
create mypage/notification index and modified mypage/notifications_controller

プロフィール一覧の通知一覧を作成しました。またnotificationのコントローラのパスを修正しました。

tetetearraio916 committed:
modified 10_notification

通知機能の修正をしました。また、アイコンのデフォルト画像をcarrierwaveで設定し直しました。

DaichiSaito commented on Gemfile:
```diff
@@ -85,3 +85,5 @@ gem 'kaminari'
 #国際化対応
 gem 'rails-i18n'
 gem 'config'
+
+gem 'jquery-rails'
```
これを入れた理由はなんでしょう？？

tetetearraio916 replied:
ドロップダウンの問題点の修正で、jqueryをyarn経由で導入していたから挙動がおかしいのかと推測して、gemで導入して修正できるか試してそのまま残していました。あまり関係なかったようなので消してしまおうと思います。

DaichiSaito commented on app/assets/javascripts/post_button_delete.js:
```diff
@@ -1,40 +0,0 @@
-//削除ボタンをクリックしたとき、処理が動く。
-$(document).on("click", ".image-preview_btn_delete", function () {
-  var append_input = $(
-    `<li class="input"><label class="upload-label"><div class="upload-label__text">ドラッグアンドドロップ<br>またはクリックしてファイルをアップロード<div class="input-area"><input class="hidden image_upload" type="file"></div></div></label></li>`
-  );
-  $ul = $("#previews");
-  $lis = $ul.find(".image-preview");
-  $input = $ul.find(".input");
-  $ul = $("#previews");
-  $li = $(this).parents(".image-preview");
-
-  //"li"ごと削除して、previewとinputを削除させる。
-  $li.remove();
-
-  // inputボタンのサイズを更新する、または追加させる
-  // まずはプレビューの数を数える。
-  $lis = $ul.find(".image-preview");
-  $label = $ul.find(".input");
-  if ($lis.length <= 4) {
-    // inputのサイズを変更
-    $("#previews li:last-child").css({
-      width: `calc(100% - (20% * ${$lis.length}))`,
-    });
-  } else if ($lis.length == 5) {
-    // inputのサイズを変更
-    $("#previews li:last-child").css({
-      width: `100%`,
-    });
-  } else if ($lis.length < 9) {
-    // inputのサイズを変更
-    $("#previews li:last-child").css({
-      width: `calc(100% - (20% * (${$lis.length} - 5 )))`,
-    });
-  } else if ($lis.length == 9) {
-    $ul.append(append_input); // 9個の時だけ、新しいinputを追加してやる
-    $("#previews li:last-child").css({
-      width: `calc(100% - (20% * (${$lis.length} - 5 )))`,
-    });
-  }
-});
```
消して大丈夫でしたか？

tetetearraio916 replied:
複数画像投稿の修正でダイソンさんのコードに直したので、大丈夫でした！消した理由は、余計なjsによってドロップダウンが上手く機能しないのではないかと思い消しました。

DaichiSaito commented on app/assets/javascripts/post_button_input.js:
```diff
@@ -1,70 +0,0 @@
-    //プレビュー完了後は、inputを非表示にさせる。これによりプレビューだけが残る。
-    $label.css("display", "none");
-    $li.removeClass("input");
-    $li.addClass("image-preview");
-    $lis = $ul.find(".image-preview");
-
-    $("#previews li").css({
-      width: `114px`,
-    });
-    //"ul"に新しい"li(inputボタン)"を追加させる。
-    if ($lis.length <= 4) {
-      $ul.append(append_input);
-      $("#previews li:last-child").css({
-        width: `calc(100% - (20% * ${$lis.length}))`,
-      });
-    } else if ($lis.length == 5) {
-      $li.addClass("image-preview");
-      $ul.append(append_input);
-      $("#previews li:last-child").css({
-        width: `100%`,
-      });
-    } else if ($lis.length <= 9) {
-      $li.addClass("image-preview");
-      $ul.append(append_input);
-      $("#previews li:last-child").css({
-        width: `calc(100% - (20% * (${$lis.length} - 5 )))`,
-      });
-    }
-
-    //inputの最後の"data-image"を取得して、input nameの番号を更新させてる。
-    $inputs.each(function (num, input) {
-      //nameの番号を更新するために、現在の番号を除去
-      $(input).removeAttr("file");
-      $(input).attr({
-        name: "post[images_attributes][" + num + "][file]",
-        id: "post_images_attributes_" + num + "_file",
-      });
-    });
-  });
-});
```
消して大丈夫でしたか？

tetetearraio916 replied:
上と理由は同じです！

DaichiSaito commented on app/controllers/mypage/notifications_controller.rb:
```diff
@@ -0,0 +1,12 @@
+class Mypage::NotificationsController < ApplicationController
+
+  def index
+    @notifications = Notification.order(id: :desc).page(params[:page]).per(10)
+  end
```
自分以外の通知も見れてしまうので一発アウトです。

tetetearraio916 replied:
完全に見落としていました。修正しました！

> current_user. notification.order(id: :desc).page(params[:page]).per(10)

tetetearraio916 committed:
delete jquery-rails from Gemfile and modified notifications_controller

Gemfileからjquery-railsのgemを特に必要がなかったので削除しました。
また、notificationsコントローラ内のindexが全ての通知が見れるようにデータを取得していたので
修正しました。

DaichiSaito commented:
コミットメッセージは最低限何をしたのかわかるようなメッセージにしましょう！
「もし自分がレビュワーだったら」を念頭に。

[![Image from Gyazo](https://i.gyazo.com/6806ca54ebcf97eaf1fcb99c1433919c.png)](https://gyazo.com/6806ca54ebcf97eaf1fcb99c1433919c)
